oak_functions_launcher/src/lookup.rs

```diff
@@ -21,55 +21,70 @@ use crate::{
 use anyhow::{anyhow, Context};
 use hashbrown::HashMap;
 use prost::Message;
-use std::{fs, path::PathBuf, vec::IntoIter};
+use std::{fs, path::PathBuf};
 use ubyte::ByteUnit;
 
-struct UpdateClient<'a> {
+struct UpdateClient<'a, I: Iterator<Item = LookupDataChunk>> {
     inner: &'a mut OakFunctionsAsyncClient<ConnectorHandle>,
-    chunks: IntoIter<LookupDataChunk>,
+    chunks: I,
 }
 
-impl UpdateClient<'_> {
+impl<I: Iterator<Item = LookupDataChunk>> UpdateClient<'_, I> {
     async fn start(&mut self) -> anyhow::Result<()> {
         // Send the current chunk first.
-        let update_response = self.send_request(UpdateAction::Start).await?;
-
+        let chunk = self.chunks.next();
+        let update_response = self.send_request(UpdateAction::Start, chunk).await?;
         if UpdateStatus::Started != update_response.update_status() {
             return Err(anyhow!("Did not receive expected update status: Started"));
         };
+        self.continue_().await
+    }
 
-        // Currently send only data which fits in two chunks.
-        // TODO(#3718): to send more chunks.
-        if self.chunks.len() > 1 {
-            return Err(anyhow!("Cannot send lookup data with more than 2 chunks"));
+    // Call continue with at least two chunks left to process.
+    async fn continue_(&mut self) -> anyhow::Result<()> {
+        // Check the next two chunks.
+        let mut next = self.chunks.next();
+        let mut next_next = self.chunks.next();
+
+        while next_next.is_some() {
```

Comment:
I'm still not sure I see the benefit of this API with two levels of lookahead, but perhaps it will be cleaned up as part of #3763?

Reply:
Yes, I want to clean-up!